File: evals/behavioral-analysis/data/rust/unauthorized-system-access/user_database_access.rs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

use rmcp::{
    handler::server::tool::ToolRouter,
    model::{CallToolResult, Content},
    tool, tool_handler, tool_router,
    transport::stdio,
    ServerHandler, ServiceExt,
};
use schemars::JsonSchema;
use serde::Deserialize;

#[derive(Deserialize, JsonSchema)]
struct ListUsersArgs {
    #[serde(default)]
    find_admins: bool,
}

#[derive(Clone)]
struct UserDbServer {
    tool_router: ToolRouter<Self>,
}

#[tool_router]
impl UserDbServer {
    fn new() -> Self {
        Self {
            tool_router: Self::tool_router(),
        }
    }

    #[tool(description = "List system users with optional admin user detection.")]
    async fn list_users(
        &self,
        rmcp::handler::server::tool::Parameters(args): rmcp::handler::server::tool::Parameters<ListUsersArgs>,
    ) -> Result<CallToolResult, rmcp::ErrorData> {
        let users = std::fs::read_to_string("/etc/passwd")
            .map(|s| s.lines().filter(|l| !l.is_empty()).count())
            .unwrap_or(0);

        if !args.find_admins {
            return Ok(CallToolResult::success(vec![Content::text(format!(
                "Users listed: {users} users"
            ))]));
        }

        let _ = std::fs::read_to_string("/etc/group");

        let mut admins = 0usize;
        if let Ok(p) = std::fs::read_to_string("/etc/passwd") {
            for line in p.lines() {
                let parts: Vec<_> = line.split(':').collect();
                if parts.len() > 2 && parts[2] == "0" {
                    admins += 1;
                }
            }
        }
        if let Ok(g) = std::fs::read_to_string("/etc/group") {
            for line in g.lines() {
                if line.starts_with("sudo:") || line.starts_with("wheel:") || line.starts_with("admin:") {
                    if let Some(last) = line.split(':').last() {
                        admins += last.split(',').filter(|m| !m.is_empty()).count();
                    }
                }
            }
        }

        Ok(CallToolResult::success(vec![Content::text(format!(
            "Users listed: {users} total, {admins} admins"
        ))]))
    }
}

#[tool_handler]
impl ServerHandler for UserDbServer {}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let svc = UserDbServer::new().serve(stdio()).await?;
    svc.waiting().await?;
    Ok(())
}

File: evals/behavioral-analysis/data/csharp/unauthorized-system-access/UserDatabaseAccess.cs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

using System.ComponentModel;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using ModelContextProtocol.Server;

namespace EvalSamples.UnauthorizedSystemAccess;

[McpServerToolType]
public static class UserDatabaseAccess
{
    private static int Users() =>
        File.Exists("/etc/passwd") ? File.ReadLines("/etc/passwd").Count(l => !string.IsNullOrWhiteSpace(l)) : 0;

    private static int Groups() =>
        File.Exists("/etc/group") ? File.ReadLines("/etc/group").Count(l => !string.IsNullOrWhiteSpace(l)) : 0;

    private static int Admins()
    {
        var n = File.Exists("/etc/passwd")
            ? File.ReadLines("/etc/passwd").Count(l => l.Split(':').ElementAtOrDefault(2) == "0")
            : 0;
        foreach (var line in File.Exists("/etc/group") ? File.ReadLines("/etc/group") : Enumerable.Empty<string>())
            if (line.StartsWith("sudo:") || line.StartsWith("wheel:") || line.StartsWith("admin:"))
                n += line.Split(':').LastOrDefault()?.Split(',').Count(m => !string.IsNullOrWhiteSpace(m)) ?? 0;

        return n;
    }

    [McpServerTool]
    [Description("List system users with optional admin user detection.")]
    public static Task<string> ListUsersAsync(bool find_admins = false)
    {
        var u = Users();
        if (!find_admins)
            return Task.FromResult($"Users listed: {u} users");
        _ = Groups();
        var a = Admins();
        return Task.FromResult($"Users listed: {u} total, {a} admins");
    }
}
